### PR TITLE
tstime/rate: add Value

### PR DESCRIPTION
**Pull request by dsnet**
Add Value, which measures the rate at which an event occurs, exponentially weighted towards recent activity.
It is guaranteed to occupy O(1) memory and operate in O(1) runtime.

### CODE REVIEW

File: tstime/rate/value.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	"math"
+	"time"
+
+	"tailscale.com/tstime/mono"
+)
+
+// Value measures the rate at which an event occurs,
+// exponentially weighted towards recent activity.
+// It is guaranteed to occupy O(1) memory and operate in O(1) runtime.
+type Value struct {
```
Comment (maisem) [inline tstime/rate/value.go:17]: add a note that it is not threadsafe

Comment (maisem) [reply tstime/rate/value.go:17]: same, but a mutex might be too expensive for my application.

Comment (dsnet) [reply tstime/rate/value.go:20]: I went just a trivial `sync.Mutex` for now. The single-threaded performance degrades from 62ns to 70ns. Around 25ns of that is just calling `mono.Now()`.

We can optimize the concurrency aspect of this in a future PR. Any ideas that I can come up greatly complicates the code, so we should review the math portion of this by itself. That itself is a large tricky enough.

File: tstime/rate/value.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	//
+	// Since our events are decaying, we can divide the number of events
+	// by the total possible accumulated value, which we determine
+	// by integrating the half-life formula from t=0 until t=∞,
+	// assuming that N₀ is 1:
+	//	∫ N(t) dt = t½ / ln(2)
+	//
+	// Recall that the integral of a curve is the area under a curve,
+	// which carries the units of the X-axis multiplied by the Y-axis.
+	// In our case this would be the units "events · seconds".
```
Comment (crawshaw) [inline tstime/rate/value.go:109]: well that's just sneaky

**Review by crawshaw — COMMENTED** [review]
there's something a bit tricky going on here I don't quite understand. do you have a reference for this algorithm I could read? N₀ is both your initial quantity when calculating valueNow, but you update it whenever Add is called and I can't convince myself that the rate computation makes sense on top of the updated value.

**Comment by dsnet** [comment]
> do you have a reference for this algorithm I could read?

Other than general resources on "half-life decay" and "exponential decay", not really. This algorithm came about because I was trying to [do some proportional decay](https://go.dev/play/p/ThceKBgvvME), and then @maisem suggested maybe instead doing something like `val*e^(-age)`, which reminded me that the suggested formula looked like atomic decay, which is what eventually led to what we got now.

There has got to be some formal computer science literature on solving this problem, but searches for things like "measure event rate algorithm" or "rate algorithm in O(1) memory and runtime" didn't really yield anything useful. You find [Stack Overflow posts](https://stackoverflow.com/questions/11731735/how-to-calculate-continuous-smooth-event-rate-based-on-event-times) that recommend decaying by 1/60th of the value, but such a solution assumes that the sample rate is fixed.

> N₀ is both your initial quantity when calculating valueNow, but you update it whenever Add is called and I can't convince myself that the rate computation makes sense on top of the updated value.

The way I reason about it is that `value` carries the units "events" and is an approximation on the total number of events still alive (i.e., has not decayed away). `r.valueNow` computes the total number of events alive right now after decaying away everything between `now` and `updated`. We add `n` since those events just happens right now and have not decayed away.

File: tstime/rate/value.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	HalfLife time.Duration
+
+	mu      sync.Mutex
+	updated mono.Time
+	value   float64 // adjusted count of events
+}
+
+// halfLife returns the half-life period in seconds.
+func (r *Value) halfLife() float64 {
+	halfLife := atomic.LoadInt64((*int64)(&r.HalfLife))
```
[the remark this answers is not in the document]
Comment (dsnet) [reply tstime/rate/value.go:45]: Ooops. This is leftover from an attempt to make this thing fully depend on atomics, but then I reduced it back to just using a `sync.Mutex` and leave optimizations for the future.

**Comment by crawshaw** [comment]
A common academic term for algorithms that process an event stream in some fixed memory window is "online." Searching for that didn't reveal anything exactly, but it did point to a popular finance technique called EWMA (exponential weighted moving average) for time series data.

https://www.itl.nist.gov/div898/handbook/pmc/section4/pmc431.htm

My first read-through of it gives me a better intuition for the updating of `value` you're doing, but it's been a long day. I'll think about it again in the morning and if it still makes sense will wave this PR on. :)

**Comment by dsnet** [comment]
Interesting. I think @andrew-d pointed at something with EWMA, but I had no idea what the acronym meant (sometimes acronyms really get in the way of understanding something).

It does appear that EWMA and the algorithm here share similarities:
* EWMA appear to be defined according to a [decay constant](https://en.wikipedia.org/wiki/Exponential_smoothing#Time_constant), which is semantically equivalent to a half-life according to the [last half-life formula variant](https://en.wikipedia.org/wiki/Half-life#Formulas_for_half-life_in_exponential_decay). I find the "half-life" representation much more intuitive as a half-life of 1s tells me that our rate will be 50% accurate within 1s, 75% accurate within 2s, 87.5% accurate within 3s, and so on. The equivalent decay constant of 0.6931 tells me nothing unless I plot it out.
* A significant difference between our `rate.Value` and an EWMA is that our `rate.Value` takes variable time into consideration, while EWMA seems to heavily assume a fixed sampling period (e.g., the daily stock price). Our rate needs to be sensible even if there is a long period of silence between events or the occurrence of events is not uniformly spread out. For example, if we want to compute the rate of how often a node disconnects, we are not in control of when that event occurs. It is simpler and more efficient to call `rate.Value.Add(1)` whenever the node disconnects, than it is to have some background ticker that checks every T period whether the node is connected or not.

**Comment by dsnet** [comment]
> Giving it a name other than `rate.Value`.

From a Go package perspective, this would introduce a naming inconsistency. At present, the package's naming convention is based on what they do (i.e., `Limiter` limits a rate, while `Value` measures a rate), rather than how they do it. Both fundamentally choose some algorithm (from among many) to do what it does, but documents which algorithm. For example, `Limiter` clearly documents that it uses the "Token bucket" algorithm.

What are your thoughts on keeping it simple like `rate.Value`, but writing a blog post and then pointing the documentation of `Value` at the blog post along with a reasonable algorithm name that we come up with?

Alternatively, we can just put `rate.Value` in a new package and thus absolve ourselves of any naming precedence.

**Comment by crawshaw** [comment]
Documenting a Google-able algorithm name in the type comments is probably good enough. The problem I'm trying to solve is "oh a rate.Value, how do I interpret this data? let me search..."

**Review by maisem — APPROVED** [review]

**Review by crawshaw — APPROVED** [review]

**Comment by dsnet** [comment]
Thanks for the reviews! I think this is the first Go package I wrote with more comments than code 😆

**Pull request merged**